File: python/send_think.py
#!/usr/bin/env python3
"""
Resonant Protocol - Python sender example
Sends Sync, Think, and Critique messages via UDP
"""

import socket
import struct
import zlib
import json
import random

HOST, PORT = "127.0.0.1", 50051

# Message types
MSG_THINK = 0
MSG_CACHE = 1
MSG_ASK = 2
MSG_SYNC = 3
MSG_CRITIQUE = 4

# Modality
MODALITY_TEXT = 0
MODALITY_IMAGE = 1

# DType
DTYPE_F16 = 0x01
DTYPE_I8 = 0x02
DTYPE_Q4 = 0x03

# Flags
FLAG_ZSTD = 1 << 0
FLAG_XCHACHA = 1 << 1
FLAG_STRONG_TAIL = 1 << 2


def build_frame(
    msg_type: int,
    stream_id: int,
    frame_seq: int,
    slices: list,
    space_hash32: int = 2451163210,
    modality: int = MODALITY_TEXT,
    flags: int = 0,
) -> bytes:
    """Build a V-Frame"""
    frame = bytearray()

    # Header
    frame += struct.pack("B", 1)  # version
    frame += struct.pack("B", msg_type)
    frame += struct.pack("<H", flags)
    frame += struct.pack("<I", stream_id)
    frame += struct.pack("<Q", frame_seq)
    frame += struct.pack("<Q", len(slices))

    # Slice lengths
    if len(slices) == 1:
        frame += struct.pack("<I", len(slices[0]["payload"]))
    else:
        for s in slices:
            frame += struct.pack("<I", len(s["payload"]))

    frame += struct.pack("<I", space_hash32)
    frame += struct.pack("B", modality)

    # Slices
    for s in slices:
        frame += struct.pack("B", s["dtype"])
        frame += struct.pack("B", len(s["shape"]))
        for dim in s["shape"]:
            frame += struct.pack("<I", dim)
        frame += s["payload"]

    # CRC32
    crc = zlib.crc32(frame) & 0xFFFFFFFF
    frame += struct.pack("<I", crc)

    return bytes(frame)


def main():
    print("🌀 Resonant Protocol - Python Sender")

    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    stream_id = random.randint(0, 0xFFFFFFFF)

    print(f"📡 Stream ID: {stream_id:#x}\n")

    # 1. Send Sync-capability
    print("1️⃣  Sending Sync-capability...")
    ping_data = json.dumps({"method": "ping", "ts": 1730616000}).encode("utf-8")

    sync_frame = build_frame(
        msg_type=MSG_SYNC,
        stream_id=stream_id,
        frame_seq=0,
        slices=[{"dtype": DTYPE_I8, "shape": [len(ping_data)], "payload": ping_data}],
    )

    sock.sendto(sync_frame, (HOST, PORT))
    print(f"✅ Sent {len(sync_frame)} bytes\n")

    # 2. Send Think with f16 latent
    print("2️⃣  Sending Think message...")
    hidden_size = 2048
    hidden_payload = b"\x00" * (hidden_size * 2)  # f16 = 2 bytes each

    think_frame = build_frame(
        msg_type=MSG_THINK,
        stream_id=stream_id,
        frame_seq=1,
        slices=[
            {"dtype": DTYPE_F16, "shape": [1, hidden_size], "payload": hidden_payload}
        ],
        flags=FLAG_ZSTD,
    )

    sock.sendto(think_frame, (HOST, PORT))
    print(f"✅ Sent {len(think_frame)} bytes")
    print(f"   Shape: [1, {hidden_size}]")
    print(f"   DType: F16\n")

    # 3. Send Critique
    print("3️⃣  Sending Critique message...")
    divergence_vec = b"\x00" * 32
    explanation = json.dumps(
        {"note": "divergence at dims [3, 17]", "magnitude": 0.042}
    ).encode("utf-8")

    critique_frame = build_frame(
        msg_type=MSG_CRITIQUE,
        stream_id=stream_id,
        frame_seq=2,
        slices=[
            {"dtype": DTYPE_F16, "shape": [16], "payload": divergence_vec},
            {"dtype": DTYPE_I8, "shape": [len(explanation)], "payload": explanation},
        ],
    )

    sock.sendto(critique_frame, (HOST, PORT))
    print(f"✅ Sent {len(critique_frame)} bytes\n")

    print("🎉 All messages sent!")
    sock.close()


if __name__ == "__main__":
    main()

File: rust/src/examples/send_think.rs
use rand::Rng;
use resonant_protocol::*;
use std::net::UdpSocket;

fn main() -> anyhow::Result<()> {
    println!("🌀 Resonant Protocol - Sending Think message");

    let sock = UdpSocket::bind("0.0.0.0:0")?;
    sock.connect("127.0.0.1:50051")?;

    let stream_id: u32 = rand::thread_rng().gen();
    println!("📡 Stream ID: {:#x}", stream_id);

    // First, send Sync-capability handshake
    println!("\n1️⃣  Sending Sync-capability handshake...");
    let capability = serde_json::json!({
        "method": "ping",
        "ts": 1730616000u64
    });

    let cap_bytes = serde_json::to_vec(&capability)?;
    let sync_frame = VFrame {
        hdr: VFrameHeader {
            version: 1,
            mtype: MsgType::Sync,
            flags: Flags::empty(),
            stream_id,
            frame_seq: 0,
            num_slices: 1,
            slice_len: vec![cap_bytes.len() as u32],
            space_hash32: 2451163210,
            modality: Modality::Text,
        },
        slices: vec![(
            SliceMeta {
                dtype: DType::I8,
                shape: vec![cap_bytes.len() as u32],
            },
            cap_bytes,
        )],
        crc32: 0,
    };

    let sync_bytes = sync_frame.encode()?;
    sock.send(&sync_bytes)?;
    println!("✅ Sent {} bytes", sync_bytes.len());

    // Send Think message with latent state
    println!("\n2️⃣  Sending Think message with f16 latent state...");

    // Create toy hidden state (4096 f16 values = 8192 bytes)
    let hidden_size = 2048;
    let payload = vec![0u8; hidden_size * 2]; // f16 = 2 bytes

    let think_frame = VFrame {
        hdr: VFrameHeader {
            version: 1,
            mtype: MsgType::Think,
            flags: Flags::ZSTD, // Enable compression
            stream_id,
            frame_seq: 1,
            num_slices: 1,
            slice_len: vec![payload.len() as u32],
            space_hash32: 2451163210,
            modality: Modality::Text,
        },
        slices: vec![(
            SliceMeta {
                dtype: DType::F16,
                shape: vec![1, hidden_size as u32],
            },
            payload,
        )],
        crc32: 0,
    };

    let think_bytes = think_frame.encode()?;
    sock.send(&think_bytes)?;
    println!("✅ Sent {} bytes", think_bytes.len());
    println!("   Shape: [1, {}]", hidden_size);
    println!("   DType: F16");

    // Send Critique message
    println!("\n3️⃣  Sending Critique message...");

    let divergence_vec = vec![0u8; 32]; // Small divergence vector
    let explanation = serde_json::json!({
        "note": "divergence at dims [3, 17]",
        "magnitude": 0.042
    });
    let explain_bytes = serde_json::to_vec(&explanation)?;

    let critique_frame = VFrame {
        hdr: VFrameHeader {
            version: 1,
            mtype: MsgType::Critique,
            flags: Flags::empty(),
            stream_id,
            frame_seq: 2,
            num_slices: 2,
            slice_len: vec![divergence_vec.len() as u32, explain_bytes.len() as u32],
            space_hash32: 2451163210,
            modality: Modality::Text,
        },
        slices: vec![
            (
                SliceMeta {
                    dtype: DType::F16,
                    shape: vec![16],
                },
                divergence_vec,
            ),
            (
                SliceMeta {
                    dtype: DType::I8,
                    shape: vec![explain_bytes.len() as u32],
                },
                explain_bytes,
            ),
        ],
        crc32: 0,
    };

    let critique_bytes = critique_frame.encode()?;
    sock.send(&critique_bytes)?;
    println!("✅ Sent {} bytes", critique_bytes.len());

    println!("\n🎉 All messages sent successfully!");
    println!("   Waiting for receiver to process...");

    Ok(())
}
